### DISCUSSION TITLE
Porting Old Mojolicious Code to Current Version

### DISCUSSION THREAD

**Discussion by ornh** [mojolicious/mojo#1891] [General]
Hi,
I am having issues trying to resurrect some old (Mojolicious version 4) code burrowed from the internet. Following are the errors I am getting:

`Mojo::Reactor::EV: I/O watcher failed: Can't locate object method "is_writing" via package "Mojo::Transaction::HTTP"`

`Mojo::Reactor::EV: Timer failed: Can't locate object method "server_close" via package "Mojo::Transaction::HTTP"`

I have been through the change log and looked at upgrading notes here: https://github.com/mojolicious/mojo/wiki/Upgrading. I see the methods have changed but there is nothing stating how these calls should be made in the latest version of Mojolicious.

Can someone please lend a hand?

Thanks

**Comment by kraih** [comment]
That's a very low level API pretty much nobody ever used. And it's been a very long time since it has been removed. So, i doubt you'll be able to find anyone who still remembers how it worked i'm afraid. Your best bet is probably to look up the version in `Changes` and then study the commits.

**Reply by ornh** [reply]
It seems very old indeed. Perhaps, rather than trying to revive the old code, you could point me in a better direction. I am trying to implement an intercepting proxy using Mojolicious. I need to be able to pause and modify requests/responses on the fly. One example would be updating Bearer tokens in request headers. I see examples of a standard HTTP proxy and a Reverse proxy but nothing for an intercepting proxy. Any suggestions?

**Reply by kraih** [reply]
That's a built-in feature now. https://docs.mojolicious.org/Mojolicious/Plugin/DefaultHelpers#proxy-start_p

**Reply by ornh** [reply]
Thank you for the link. However, this seems to be for building out requests by hand. How would I accept requests from a browser, modify the request, send the request to the server and then send the response to the browser?

**Comment by kraih** [comment]
Considering that there's also a very high chance that those methods have been used wrong (they are not really something a Mojolicious app would use), maybe you might even be better off just trying to fix the code.

**Comment by ornh** [comment]
Hi kraih,
After a good bit of searching, I finally found the examples page and the connect-proxy example of yours [https://github.com/mojolicious/mojo/blob/main/examples/connect-proxy.pl](url). This looks to be very close to what I need. I can see that $buffer has the full request and can be manipulated prior to the request being sent. However, I don't see any response headers, only the body returned in $chunk. Is there a way this code could be modified to use Mojo::UserAgent and parse the requests/responses? Or at least get to the response headers in the current code. Sorry if this is a very noob question, I am fairly new to OOP and this is my first time using Mojo.

Your help is greatly appreciated,
Thank you

**Reply by ornh** [reply]
I am just now getting back to this. Any help???
